### ISSUE TITLE
Baselines outperform PAM-CDR

### ISSUE THREAD

**Issue by PascalIversen** [catly/PAM-CDR#1]
Hello, 

Thank you for publishing the code! :)

I found that the following simple baselines (Linreg, Random forest) all outperform PAM-CDR without using any cell line features (just one hot encoding). Could you try this and confirm? Could you also publish the input data you used (rna expression etc), so I can run your model?

Thank you very much.

```
================================================================================
VALIDATION RESULTS
================================================================================

Training RF-500 (drug+cell)...
  Val                            | AUC: 0.9703 | AUPR: 0.9834 | Acc: 0.9082
  Train                          | AUC: 0.9699 | AUPR: 0.9833 | Acc: 0.9077

Training RF-1000 (drug+cell)...
  Val                            | AUC: 0.9700 | AUPR: 0.9832 | Acc: 0.9078
  Train                          | AUC: 0.9695 | AUPR: 0.9831 | Acc: 0.9077

Training GBT-500 (drug+cell)...
  Val                            | AUC: 0.9701 | AUPR: 0.9831 | Acc: 0.9117
  Train                          | AUC: 0.9864 | AUPR: 0.9919 | Acc: 0.9263

Training LogReg (drug+cell)...
  Val                            | AUC: 0.9738 | AUPR: 0.9855 | Acc: 0.9132
  Train                          | AUC: 0.9776 | AUPR: 0.9876 | Acc: 0.9219

Training RF-500 (drug only)...
  Val                            | AUC: 0.9704 | AUPR: 0.9834 | Acc: 0.9078
  Train                          | AUC: 0.9699 | AUPR: 0.9833 | Acc: 0.9077

================================================================================
TEST RESULTS
================================================================================
  RF-500 (drug+cell)             | AUC: 0.9743 | AUPR: 0.9857 | Acc: 0.9152
  RF-1000 (drug+cell)            | AUC: 0.9738 | AUPR: 0.9853 | Acc: 0.9156
  GBT-500 (drug+cell)            | AUC: 0.9740 | AUPR: 0.9853 | Acc: 0.9186
  LogReg (drug+cell) ** best     | AUC: 0.9780 | AUPR: 0.9879 | Acc: 0.9225
  RF-500 (drug only)             | AUC: 0.9747 | AUPR: 0.9857 | Acc: 0.9156
```

```python
"""
Random Forest baseline for PAM-CDR cancer drug response prediction.
Works directly from the repo's CSV files

Features:
  - Drug: Morgan fingerprints (2048-bit) from SMILES
  - Cell line: one-hot encoding

Uses the same split logic as the deep model (80/10/10, seed 42).

Usage:
    pip install rdkit scikit-learn pandas numpy
    python baseline_rf.py
"""

import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score, precision_recall_curve, auc, accuracy_score
from rdkit import Chem
from rdkit.Chem import AllChem
import warnings
warnings.filterwarnings("ignore")


def smiles_to_morgan(smiles, radius=2, n_bits=2048):
    mol = Chem.MolFromSmiles(smiles)
    if mol is None:
        return np.zeros(n_bits)
    fp = AllChem.GetMorganFingerprintAsBitVect(mol, radius, nBits=n_bits)
    return np.array(fp)


def evaluate(y_true, y_prob, name=""):
    auc_score = roc_auc_score(y_true, y_prob)
    precision, recall, _ = precision_recall_curve(y_true, y_prob)
    aupr = auc(recall, precision)
    acc = accuracy_score(y_true, (y_prob >= 0.5).astype(int))
    print(f"  {name:30s} | AUC: {auc_score:.4f} | AUPR: {aupr:.4f} | Acc: {acc:.4f}")
    return auc_score, aupr, acc


def split_data(data):
    n = len(data)
    train = data[:int(n * 0.8)]
    val = data[int(n * 0.8):int(n * 0.9)]
    test = data[int(n * 0.9):]
    return train, val, test


def main():
    # ── Load CSVs ──
    print("Loading data...")
    sensitive = pd.read_csv('cell_data/sensitivity_12571.csv')
    resistant = pd.read_csv('cell_data/resistant_7809.csv')
    drug_smiles = pd.read_csv('drug_process/drugandsmile.csv')

    print(f"  Sensitive: {len(sensitive)} | Resistant: {len(resistant)}")

    # ── Build drug fingerprints ──
    print("Computing Morgan fingerprints...")
    smiles_dict = dict(zip(drug_smiles['drug'], drug_smiles['smiles']))
    all_drugs = set(sensitive['drug'].unique()) | set(resistant['drug'].unique())

    drug_fp = {}
    failed = 0
    for drug in all_drugs:
        if drug in smiles_dict:
            drug_fp[drug] = smiles_to_morgan(smiles_dict[drug])
        else:
            drug_fp[drug] = np.zeros(2048)
            failed += 1
    print(f"  {len(drug_fp)} drugs fingerprinted ({failed} missing SMILES)")

    # ── Encode cell lines ──
    all_cells = sorted(set(sensitive['cell_lines'].unique()) | set(resistant['cell_lines'].unique()))
    cell_to_idx = {c: i for i, c in enumerate(all_cells)}
    n_cells = len(all_cells)
    print(f"  {n_cells} unique cell lines")

    # ── Replicate the original split logic ──
    sens_arr = sensitive[['cell_lines', 'drug', 'label']].values
    res_arr = resistant[['cell_lines', 'drug', 'label']].values

    np.random.seed(42)
    np.random.shuffle(sens_arr)
    np.random.shuffle(res_arr)

    s_train, s_val, s_test = split_data(sens_arr)
    r_train, r_val, r_test = split_data(res_arr)

    train_data = np.vstack((s_train, r_train))
    val_data = np.vstack((s_val, r_val))
    test_data = np.vstack((s_test, r_test))

    print(f"  Train: {len(train_data)} | Val: {len(val_data)} | Test: {len(test_data)}")

    #  Build feature matrices 
    def build_features(data):
        X_fp = np.array([drug_fp.get(row[1], np.zeros(2048)) for row in data])
        X_cell = np.zeros((len(data), n_cells))
        for i, row in enumerate(data):
            idx = cell_to_idx.get(row[0])
            if idx is not None:
                X_cell[i, idx] = 1.0
        X = np.hstack([X_fp, X_cell])
        y = data[:, 2].astype(int)
        return X, y

    print("Building features...")
    X_train, y_train = build_features(train_data)
    X_val, y_val = build_features(val_data)
    X_test, y_test = build_features(test_data)
    print(f"  Feature dim: {X_train.shape[1]} (2048 Morgan + {n_cells} cell one-hot)")
    print(f"  Train label balance: {y_train.mean():.3f} sensitive")
    print()

    #  Drug-only features for ablation 
    def build_drug_only(data):
        X_fp = np.array([drug_fp.get(row[1], np.zeros(2048)) for row in data])
        y = data[:, 2].astype(int)
        return X_fp, y

    X_train_drug, _ = build_drug_only(train_data)
    X_val_drug, _ = build_drug_only(val_data)
    X_test_drug, _ = build_drug_only(test_data)

    # Models
    configs = [
        ("RF-500 (drug+cell)", RandomForestClassifier(
            n_estimators=500, max_depth=None, min_samples_leaf=3,
            n_jobs=-1, random_state=42), True),
        ("RF-1000 (drug+cell)", RandomForestClassifier(
            n_estimators=1000, max_depth=30, min_samples_leaf=3,
            n_jobs=-1, random_state=42), True),
        ("GBT-500 (drug+cell)", GradientBoostingClassifier(
            n_estimators=500, max_depth=5, learning_rate=0.05,
            subsample=0.8, random_state=42), True),
        ("LogReg (drug+cell)", LogisticRegression(
            max_iter=1000, C=1.0, random_state=42), True),
        ("RF-500 (drug only)", RandomForestClassifier(
            n_estimators=500, max_depth=None, min_samples_leaf=3,
            n_jobs=-1, random_state=42), False),
    ]

    print("=" * 80)
    print("VALIDATION RESULTS")
    print("=" * 80)

    best_val_auc = 0
    best_name = None
    results = {}

    for name, model, use_cell in configs:
        Xtr = X_train if use_cell else X_train_drug
        Xv = X_val if use_cell else X_val_drug
        print(f"\nTraining {name}...")
        model.fit(Xtr, y_train)

        y_prob_val = model.predict_proba(Xv)[:, 1]
        val_auc, val_aupr, val_acc = evaluate(y_val, y_prob_val, "Val")

        y_prob_train = model.predict_proba(Xtr)[:, 1]
        evaluate(y_train, y_prob_train, "Train")

        results[name] = (model, use_cell)
        if val_auc > best_val_auc:
            best_val_auc = val_auc
            best_name = name

    # ── Test all models ──
    print("\n" + "=" * 80)
    print("TEST RESULTS")
    print("=" * 80)

    for name, (model, use_cell) in results.items():
        Xt = X_test if use_cell else X_test_drug
        y_prob_test = model.predict_proba(Xt)[:, 1]
        marker = " ** best" if name == best_name else ""
        evaluate(y_test, y_prob_test, f"{name}{marker}")


if __name__ == "__main__":
    main()
```

**Comment by catly** [comment]
Dear Pascal, 
Thank you for your attention. 
We would like to clarify one key point that may have been overlooked.
Our model is biologically grounded, as it leverages cell line features. Therefore, it is designed to generalize to previously unseen cell lines.
In contrast, your implementation uses cell line IDs as input features. Under this setting, the model is likely to learn identity-specific patterns rather than biologically meaningful representations, which may limit its ability to generalize to unseen cell lines.

**Comment by PascalIversen** [comment]
Thanks for the response.

I understand the argument about biological grounding, but I don't think it addresses the core issue:


The dataset is dominated by drugs with constant labels. 173 out of 263 drugs (65.8%) have a 100% sensitivity or resistance rate across all cell lines, meaning every cell line responds identically to those drugs. At a 90% purity threshold, this rises to 210/263 drugs covering 70% of all samples.  

Anti-cancer drugs have a very different overall potency. You did not properly normalize the labels to create a meaningful prediction task here.

The drug-only baseline (no cell line features at all) already outperforms your model (AUC 0.9747 vs your reported results). This baseline uses only Morgan fingerprints. If drug features alone beat a model that uses drug + RNA + CNV, it suggests the task under this evaluation setup is largely solvable from drug identity alone.


It is not supported that the performance of your model are from learning biology.
Your train/test split leaks both drugs and cell lines across splits. The same drug-cell line pairs aren't shared, but the same drugs and same cell lines appear in both train and test. This means any model, including yours, can exploit drug-level or cell-level memorization.

The claim of "generalizing to unseen cell lines" isn't actually tested by your evaluation protocol.

To assess whether your model actually learns drug-cell interactions, could you report per-drug AUC (AUC computed within each drug, averaged across drugs) for a leave-cell-out split? This controls for drug-level priors and reveals whether the model meaningfully uses cell line features to distinguish response within the same drug.


I'd also appreciate access to the preprocessed data so I can reproduce your results directly and can try a simple baseline with RNA features.


Thanks a lot for your help!

Pascal

**Comment by PascalIversen** [comment]
<img width="2379" height="973" alt="Image" src="https://github.com/user-attachments/assets/ebc52331-4b6a-43d6-b955-9892561fa4db" />

Here is a viz of the problem.

**Comment by catly** [comment]
Sorry for the late reply.
Thank you for this careful and constructive clarification. The Cmax-based binarization strategy adopted in our study has been widely used in previous GDSC-based drug response studies [1–3].

We are now evaluating our model on larger scale datasets to further assess its robustness and generalization ability. We also plan to adopt more stringent data-splitting strategies and evaluation metrics to ensure a more comprehensive assessment of model performance. We sincerely appreciate your valuable suggestion.


[1] Iorio F, Knijnenburg TA, Vis DJ, Bignell GR, Menden MP, Schubert M, Aben N, Gonçalves E, Barthorpe S, Lightfoot H, Cokelaer T, Greninger P, van Dyk E, Chang H, de Silva H, Heyn H, Deng X, Egan RK, Liu Q, Mironenko T, Mitropoulos X, Richardson L, Wang J, Zhang T, Moran S, Sayols S, Soleimani M, Tamborero D, Lopez-Bigas N, Ross-Macdonald P, Esteller M, Gray NS, Haber DA, Stratton MR, Benes CH, Wessels LFA, Saez-Rodriguez J, McDermott U, Garnett MJ. A Landscape of Pharmacogenomic Interactions in Cancer. Cell. 2016 Jul 28;166(3):740-754.
[2] Liu X, Song C, Huang F, Fu H, Xiao W, Zhang W. GraphCDR: a graph neural network method with contrastive learning for cancer drug response prediction. Brief Bioinform. 2022 Jan 17;23(1):bbab457.
[3] Choi J, Park S, Ahn J. RefDNN: a reference drug based neural network for more accurate prediction of anticancer drug resistance. Sci Rep. 2020 Feb 5;10(1):1861.
